File: src/mcts.py
"""Monte Carlo Tree Search for AlphaZero."""

from __future__ import annotations

import math
from typing import TYPE_CHECKING

import numpy as np
import torch

from .game import ConnectFour, COLS

if TYPE_CHECKING:
    from .model import AlphaZeroNet


class MCTSNode:
    """A node in the MCTS tree."""

    __slots__ = (
        "game",
        "parent",
        "action",
        "children",
        "visit_count",
        "total_value",
        "prior",
        "is_expanded",
    )

    def __init__(
        self,
        game: ConnectFour,
        parent: MCTSNode | None = None,
        action: int | None = None,
        prior: float = 0.0,
    ) -> None:
        self.game = game
        self.parent = parent
        self.action = action
        self.prior = prior
        self.children: dict[int, MCTSNode] = {}
        self.visit_count: int = 0
        self.total_value: float = 0.0
        self.is_expanded: bool = False

    @property
    def q_value(self) -> float:
        """Mean action value."""
        if self.visit_count == 0:
            return 0.0
        return self.total_value / self.visit_count

    def ucb_score(self, c_puct: float = 1.5) -> float:
        """Upper Confidence Bound score using PUCT formula.

        Q-value is negated because each child stores value from its own
        current player's perspective (the opponent of the selecting parent).
        The parent wants to maximize its own value = minimize child's value.
        """
        parent_visits = self.parent.visit_count if self.parent else 1
        exploration = c_puct * self.prior * math.sqrt(parent_visits) / (1 + self.visit_count)
        return -self.q_value + exploration

    def select_child(self, c_puct: float = 1.5) -> MCTSNode:
        """Select child with highest UCB score."""
        return max(self.children.values(), key=lambda c: c.ucb_score(c_puct))

    def expand(self, policy: np.ndarray) -> None:
        """Expand node using policy from neural network."""
        self.is_expanded = True
        valid_moves = self.game.get_valid_moves()

        for col in range(COLS):
            if valid_moves[col]:
                child_game = self.game.copy()
                child_game.make_move(col)
                self.children[col] = MCTSNode(
                    game=child_game,
                    parent=self,
                    action=col,
                    prior=policy[col],
                )

    def backpropagate(self, value: float) -> None:
        """Backpropagate value up the tree.

        Value is from the perspective of the player who just moved
        at this node's PARENT. We negate at each level because
        alternating players.
        """
        node: MCTSNode | None = self
        while node is not None:
            node.visit_count += 1
            # value is from perspective of the player who is about to move
            # at this node. Negate because parent's player is the opponent.
            node.total_value += value
            value = -value
            node = node.parent


class MCTS:
    """Monte Carlo Tree Search using AlphaZero approach."""

    def __init__(
        self,
        model: AlphaZeroNet,
        num_simulations: int = 200,
        c_puct: float = 1.5,
        dirichlet_alpha: float = 1.0,
        dirichlet_epsilon: float = 0.25,
        device: torch.device | None = None,
    ) -> None:
        self.model = model
        self.num_simulations = num_simulations
        self.c_puct = c_puct
        self.dirichlet_alpha = dirichlet_alpha
        self.dirichlet_epsilon = dirichlet_epsilon
        self.device = device or torch.device("cpu")

    @torch.no_grad()
    def _evaluate(self, game: ConnectFour) -> tuple[np.ndarray, float]:
        """Use neural network to evaluate a position.

        Returns:
            policy: Probability distribution over moves (numpy array, length 7).
            value: Position evaluation from current player's perspective.
        """
        state = torch.from_numpy(game.encode()).to(self.device)
        valid_moves = torch.from_numpy(game.get_valid_moves()).to(self.device)
        policy, value = self.model.predict(state, valid_moves)
        return policy.cpu().numpy(), value.item()

    def search(
        self, game: ConnectFour, add_noise: bool = True
    ) -> np.ndarray:
        """Run MCTS from the given game state.

        Args:
            game: Current game state.
            add_noise: Whether to add Dirichlet noise at root (for training).

        Returns:
            Visit count distribution over actions (normalized).
        """
        root = MCTSNode(game)

        # Evaluate and expand root
        policy, _ = self._evaluate(game)
        root.expand(policy)

        # Add Dirichlet noise at root for exploration
        if add_noise and root.children:
            noise = np.random.dirichlet(
                [self.dirichlet_alpha] * len(root.children)
            )
            for i, child in enumerate(root.children.values()):
                child.prior = (
                    (1 - self.dirichlet_epsilon) * child.prior
                    + self.dirichlet_epsilon * noise[i]
                )

        # Run simulations
        for _ in range(self.num_simulations):
            node = root

            # Selection: traverse tree using UCB
            while node.is_expanded and node.children:
                node = node.select_child(self.c_puct)

            # Check if terminal
            result = node.game.get_result()
            if result is not None:
                # Terminal node: backpropagate actual result
                node.backpropagate(result)
                continue

            # Expansion: evaluate with neural network and expand
            policy, value = self._evaluate(node.game)
            node.expand(policy)

            # Backpropagation
            node.backpropagate(value)

        # Construct action probabilities from visit counts
        visit_counts = np.zeros(COLS, dtype=np.float32)
        for action, child in root.children.items():
            visit_counts[action] = child.visit_count

        return visit_counts

    def get_action_probs(
        self,
        game: ConnectFour,
        temperature: float = 1.0,
        add_noise: bool = True,
    ) -> tuple[np.ndarray, np.ndarray]:
        """Get action probabilities from MCTS.

        Args:
            game: Current game state.
            temperature: Temperature for action selection.
                1.0 = proportional to visit counts.
                0 = greedy (pick best move).
            add_noise: Whether to add exploration noise.

        Returns:
            action_probs: Probability distribution over actions.
            visit_counts: Raw visit counts.
        """
        visit_counts = self.search(game, add_noise=add_noise)

        if temperature == 0:
            # Greedy: pick the action with most visits
            action_probs = np.zeros(COLS, dtype=np.float32)
            best_action = np.argmax(visit_counts)
            action_probs[best_action] = 1.0
        else:
            # Temperature-scaled
            if visit_counts.sum() == 0:
                # Fallback: uniform over valid moves
                valid = game.get_valid_moves().astype(np.float32)
                action_probs = valid / valid.sum()
            else:
                counts = visit_counts ** (1.0 / temperature)
                action_probs = counts / counts.sum()

        return action_probs, visit_counts

    def get_move_values(self, game: ConnectFour) -> dict[int, float]:
        """Get value estimates for each valid move (for display hints).

        Returns dict mapping column -> estimated win probability (0-100%).
        """
        # Run a full search and build the tree
        root = MCTSNode(game)
        policy, _ = self._evaluate(game)
        root.expand(policy)

        for _ in range(self.num_simulations):
            node = root
            while node.is_expanded and node.children:
                node = node.select_child(self.c_puct)
            result = node.game.get_result()
            if result is not None:
                node.backpropagate(result)
                continue
            p, v = self._evaluate(node.game)
            node.expand(p)
            node.backpropagate(v)

        values = {}
        for action, child in root.children.items():
            # Child Q-value is from child's perspective (opponent).
            # Negate to get value from current player's perspective.
            win_prob = (-child.q_value + 1) / 2 * 100  # Convert [-1,1] to [0,100]%
            values[action] = win_prob

        return values

File: rust/src/mcts.rs
/// Arena-based MCTS tree for high-performance tree search.
///
/// Nodes are stored in a contiguous Vec (arena allocation) and referenced
/// by index, avoiding heap fragmentation and borrow checker issues.
///
/// Supports virtual losses for parallel leaf collection: multiple leaves
/// can be selected simultaneously without duplicating paths.

use crate::game::{ConnectFour, COLS};
use rand_distr::{Dirichlet, Distribution};

pub struct MCTSNode {
    pub game: ConnectFour,
    pub children: Vec<(u8, usize)>, // (action, child_index)
    pub parent: Option<usize>,
    pub visit_count: u32,
    pub total_value: f64,
    pub prior: f32,
    pub is_expanded: bool,
    /// Virtual losses currently applied (for parallel leaf selection).
    pub virtual_losses: u32,
}

pub struct MCTSTree {
    pub nodes: Vec<MCTSNode>,
}

impl MCTSTree {
    /// Create a new tree with the given game state as root.
    pub fn new(game: ConnectFour) -> Self {
        let root = MCTSNode {
            game,
            children: Vec::new(),
            parent: None,
            visit_count: 0,
            total_value: 0.0,
            prior: 0.0,
            is_expanded: false,
            virtual_losses: 0,
        };
        MCTSTree { nodes: vec![root] }
    }

    #[inline]
    pub fn root(&self) -> usize {
        0
    }

    /// Select a leaf node by traversing from root using UCB.
    /// Virtual losses are respected during selection.
    pub fn select_leaf(&self, root: usize, c_puct: f32) -> usize {
        let mut idx = root;
        loop {
            let node = &self.nodes[idx];
            if !node.is_expanded || node.children.is_empty() {
                return idx;
            }
            idx = self.best_child(idx, c_puct);
        }
    }

    /// Select child with highest UCB score, accounting for virtual losses.
    fn best_child(&self, parent_idx: usize, c_puct: f32) -> usize {
        let parent = &self.nodes[parent_idx];
        let effective_parent_visits = parent.visit_count + parent.virtual_losses;
        let sqrt_parent = (effective_parent_visits.max(1) as f64).sqrt();

        let mut best_idx = 0;
        let mut best_score = f64::NEG_INFINITY;

        for &(_, child_idx) in &parent.children {
            let child = &self.nodes[child_idx];
            let effective_visits = child.visit_count + child.virtual_losses;
            let q = if effective_visits == 0 {
                0.0
            } else {
                // Virtual losses add negative value (pessimistic), discouraging
                // re-selection of in-flight leaves.
                (child.total_value - child.virtual_losses as f64)
                    / effective_visits as f64
            };
            let exploration =
                c_puct as f64 * child.prior as f64 * sqrt_parent / (1.0 + effective_visits as f64);
            // Negate Q because child stores value from its own perspective
            let score = -q + exploration;
            if score > best_score {
                best_score = score;
                best_idx = child_idx;
            }
        }
        best_idx
    }

    /// Apply virtual loss along the path from node to root.
    /// This makes the path look worse, discouraging parallel selection
    /// of the same leaf.
    pub fn apply_virtual_loss(&mut self, node_idx: usize) {
        let mut idx = Some(node_idx);
        while let Some(i) = idx {
            self.nodes[i].virtual_losses += 1;
            idx = self.nodes[i].parent;
        }
    }

    /// Revert virtual loss along the path from node to root.
    pub fn revert_virtual_loss(&mut self, node_idx: usize) {
        let mut idx = Some(node_idx);
        while let Some(i) = idx {
            self.nodes[i].virtual_losses = self.nodes[i].virtual_losses.saturating_sub(1);
            idx = self.nodes[i].parent;
        }
    }

    /// Select up to `k` distinct leaves using virtual losses.
    /// Returns a list of leaf node indices.
    /// Terminal leaves are NOT included; they are backpropagated immediately.
    pub fn select_multiple_leaves(
        &mut self,
        root: usize,
        c_puct: f32,
        k: usize,
    ) -> (Vec<usize>, u32) {
        let mut leaves = Vec::with_capacity(k);
        let mut terminal_count = 0u32;

        for _ in 0..k {
            let leaf = self.select_leaf(root, c_puct);

            let result = self.nodes[leaf].game.get_result();
            if let Some(val) = result {
                // Terminal node: backpropagate immediately, no virtual loss needed
                self.backpropagate(leaf, val);
                terminal_count += 1;
            } else {
                // Apply virtual loss to discourage re-selection
                self.apply_virtual_loss(leaf);
                leaves.push(leaf);
            }
        }

        (leaves, terminal_count)
    }

    /// Expand a node: create child nodes for each valid move.
    pub fn expand(&mut self, node_idx: usize, policy: &[f32]) {
        let game = self.nodes[node_idx].game.clone();
        let valid = game.get_valid_moves();

        let mut children = Vec::new();
        for col in 0..COLS {
            if valid[col] {
                let mut child_game = game.clone();
                child_game.make_move(col as u8);
                let child_idx = self.nodes.len();
                self.nodes.push(MCTSNode {
                    game: child_game,
                    children: Vec::new(),
                    parent: Some(node_idx),
                    visit_count: 0,
                    total_value: 0.0,
                    prior: policy[col],
                    is_expanded: false,
                    virtual_losses: 0,
                });
                children.push((col as u8, child_idx));
            }
        }
        self.nodes[node_idx].children = children;
        self.nodes[node_idx].is_expanded = true;
    }

    /// Backpropagate value up the tree, negating at each level.
    pub fn backpropagate(&mut self, node_idx: usize, mut value: f64) {
        let mut idx = Some(node_idx);
        while let Some(i) = idx {
            self.nodes[i].visit_count += 1;
            self.nodes[i].total_value += value;
            value = -value;
            idx = self.nodes[i].parent;
        }
    }

    /// Add Dirichlet noise to root's children priors.
    pub fn add_dirichlet_noise(&mut self, node_idx: usize, alpha: f64, epsilon: f64) {
        let num_children = self.nodes[node_idx].children.len();
        if num_children < 2 {
            return;
        }

        let dir = Dirichlet::new(&vec![alpha; num_children]).unwrap();
        let noise: Vec<f64> = dir.sample(&mut rand::thread_rng());

        let children: Vec<(u8, usize)> = self.nodes[node_idx].children.clone();
        for (i, &(_, child_idx)) in children.iter().enumerate() {
            let prior = self.nodes[child_idx].prior as f64;
            self.nodes[child_idx].prior = ((1.0 - epsilon) * prior + epsilon * noise[i]) as f32;
        }
    }

    /// Get visit counts for root's children as a [f32; COLS] array.
    pub fn get_visit_counts(&self, node_idx: usize) -> [f32; COLS] {
        let mut counts = [0.0f32; COLS];
        for &(action, child_idx) in &self.nodes[node_idx].children {
            counts[action as usize] = self.nodes[child_idx].visit_count as f32;
        }
        counts
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_new_tree() {
        let game = ConnectFour::new();
        let tree = MCTSTree::new(game);
        assert_eq!(tree.nodes.len(), 1);
        assert!(!tree.nodes[0].is_expanded);
        assert_eq!(tree.nodes[0].visit_count, 0);
    }

    #[test]
    fn test_expand_root() {
        let game = ConnectFour::new();
        let mut tree = MCTSTree::new(game);
        let policy = [1.0 / 7.0; COLS];
        tree.expand(0, &policy);

        assert!(tree.nodes[0].is_expanded);
        assert_eq!(tree.nodes[0].children.len(), 7); // All 7 columns valid
        assert_eq!(tree.nodes.len(), 8); // root + 7 children
    }

    #[test]
    fn test_select_leaf_unexpanded() {
        let game = ConnectFour::new();
        let tree = MCTSTree::new(game);
        assert_eq!(tree.select_leaf(0, 1.5), 0); // Root is leaf
    }

    #[test]
    fn test_select_leaf_after_expand() {
        let game = ConnectFour::new();
        let mut tree = MCTSTree::new(game);
        let policy = [1.0 / 7.0; COLS];
        tree.expand(0, &policy);

        // Should select one of the children (all equal prior, 0 visits)
        let leaf = tree.select_leaf(0, 1.5);
        assert!(leaf >= 1 && leaf <= 7);
    }

    #[test]
    fn test_backpropagate() {
        let game = ConnectFour::new();
        let mut tree = MCTSTree::new(game);
        let policy = [1.0 / 7.0; COLS];
        tree.expand(0, &policy);

        let child_idx = tree.nodes[0].children[0].1;
        tree.backpropagate(child_idx, 0.5);

        assert_eq!(tree.nodes[child_idx].visit_count, 1);
        assert!((tree.nodes[child_idx].total_value - 0.5).abs() < 1e-6);
        // Parent gets negated value
        assert_eq!(tree.nodes[0].visit_count, 1);
        assert!((tree.nodes[0].total_value - (-0.5)).abs() < 1e-6);
    }

    #[test]
    fn test_visit_counts() {
        let game = ConnectFour::new();
        let mut tree = MCTSTree::new(game);
        let policy = [1.0 / 7.0; COLS];
        tree.expand(0, &policy);

        // Backprop through first child a few times
        let child_idx = tree.nodes[0].children[0].1;
        tree.backpropagate(child_idx, 1.0);
        tree.backpropagate(child_idx, -1.0);

        let counts = tree.get_visit_counts(0);
        let action = tree.nodes[0].children[0].0;
        assert_eq!(counts[action as usize], 2.0);
    }

    #[test]
    fn test_dirichlet_noise() {
        let game = ConnectFour::new();
        let mut tree = MCTSTree::new(game);
        let policy = [1.0 / 7.0; COLS];
        tree.expand(0, &policy);

        // Get priors before noise
        let before: Vec<f32> = tree.nodes[0]
            .children
            .iter()
            .map(|&(_, idx)| tree.nodes[idx].prior)
            .collect();

        tree.add_dirichlet_noise(0, 1.0, 0.25);

        // Priors should have changed
        let after: Vec<f32> = tree.nodes[0]
            .children
            .iter()
            .map(|&(_, idx)| tree.nodes[idx].prior)
            .collect();

        assert_ne!(before, after);
    }

    #[test]
    fn test_virtual_loss_apply_revert() {
        let game = ConnectFour::new();
        let mut tree = MCTSTree::new(game);
        let policy = [1.0 / 7.0; COLS];
        tree.expand(0, &policy);

        let child_idx = tree.nodes[0].children[0].1;

        // Apply virtual loss
        tree.apply_virtual_loss(child_idx);
        assert_eq!(tree.nodes[child_idx].virtual_losses, 1);
        assert_eq!(tree.nodes[0].virtual_losses, 1); // parent too

        // Apply again
        tree.apply_virtual_loss(child_idx);
        assert_eq!(tree.nodes[child_idx].virtual_losses, 2);
        assert_eq!(tree.nodes[0].virtual_losses, 2);

        // Revert once
        tree.revert_virtual_loss(child_idx);
        assert_eq!(tree.nodes[child_idx].virtual_losses, 1);
        assert_eq!(tree.nodes[0].virtual_losses, 1);

        // Revert again
        tree.revert_virtual_loss(child_idx);
        assert_eq!(tree.nodes[child_idx].virtual_losses, 0);
        assert_eq!(tree.nodes[0].virtual_losses, 0);
    }

    #[test]
    fn test_select_multiple_leaves_diversity() {
        let game = ConnectFour::new();
        let mut tree = MCTSTree::new(game);
        let policy = [1.0 / 7.0; COLS];
        tree.expand(0, &policy);

        // Expand all children so we can go deeper
        for i in 0..7 {
            let child_idx = tree.nodes[0].children[i].1;
            tree.expand(child_idx, &policy);
        }

        // Select 4 leaves — virtual losses should push to different branches
        let (leaves, _terminal) = tree.select_multiple_leaves(0, 1.5, 4);
        assert_eq!(leaves.len(), 4);

        // At least some leaves should be different (virtual losses diversify)
        // With equal priors and no visits, all children look the same initially,
        // but virtual losses should push subsequent selections elsewhere.
        let unique: std::collections::HashSet<usize> = leaves.iter().cloned().collect();
        assert!(unique.len() >= 2, "Expected diverse leaf selection, got {:?}", leaves);

        // Clean up virtual losses
        for &leaf in &leaves {
            tree.revert_virtual_loss(leaf);
        }

        // Virtual losses should be back to 0
        for node in &tree.nodes {
            assert_eq!(node.virtual_losses, 0);
        }
    }
}
